### PR TITLE
Further improve diagnostics for expressions in pattern position

### PR DESCRIPTION
**Pull request by ShE3py**
Follow-up of #118625, see #121697.

```rs
fn main() {
    match 'b' {
        y.0.0.1.z().f()? as u32 => {},
    }
}
```
Before:
```
error: expected one of `=>`, `@`, `if`, or `|`, found `.`
 --> src/main.rs:3:10
  |
3 |         y.0.0.1.z().f()? as u32 => {},
  |          ^ expected one of `=>`, `@`, `if`, or `|`
```
After:
```
error: expected a pattern, found an expression
 --> src/main.rs:3:9
  |
3 |         y.0.0.1.z().f()? as u32 => {},
  |         ^^^^^^^^^^^^^^^^^^^^^^^ arbitrary expressions are not allowed in patterns
  |
help: consider moving the expression to a match arm guard
  |
3 |         val if val == y.0.0.1.z().f()? as u32 => {},
  |         ~~~ +++++++++++++++++++++++++++++++++
help: consider extracting the expression into a `const`
  |
2 +     const VAL: /* Type */ = y.0.0.1.z().f()? as u32;
3 ~     match 'b' {
4 ~         VAL => {},
  |
help: consider wrapping the expression in an inline `const` (requires `#![feature(inline_const_pat)]`)
  |
3 |         const { y.0.0.1.z().f()? as u32 } => {},
  |         +++++++                         +
```

---

r? fmease
@rustbot label +A-diagnostics +A-parser +A-patterns +C-enhancement

### CODE REVIEW

File: compiler/rustc_errors/src/lib.rs
Comment (ShE3py) [inline compiler/rustc_errors/src/lib.rs:815]: Not sure if this makes sense; I needed something like that in `parse_pat_tuple_or_parens()` to avoid emitting too many diagnostics since each parenthesis could have a trailing expression:

```
(1 + 2) * 3
   ~~~  ~~~
```
```
(1 + 2) * 3
 ~~~~~

(1 + 2) * 3
~~~~~~~~~~~
```

**Comment by fmease** [comment]
Currently vacationing for two more days, then I'm gonna review this PR.

**Comment by ShE3py** [comment]
Squashed the commits, I still have the `let expr` to do (e.g. `let buf[0] = expr // remove this let`) and some comments to update.

File: compiler/rustc_parse/src/parser/item.rs
Comment (ShE3py) [inline compiler/rustc_parse/src/parser/item.rs:2511]: This should be a loop (as to handle `&mut &mut PatKind::Err`), but I'm not sure if it isn't better to just returns `PatKind::Err` instead of `PatKind::Ref(PatKind::Err, _)` when parsing *ReferencePattern*s if the subpat isn't valid, at the cost of loosing information.

Same with `PatKind::Range`, when one of its bounds is an `ExprKind::Err`.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:379]: Note: This also permits integer literals with suffixes (e.g., `0i32`) contrary to your custom `is_two_tuple_indexes`. I'm fine either way (including/excluding suffixed literals) as long as it's consistent.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:364]:
```suggestion
        let is_two_tuple_indexes = |this: &Self, token: &Token| -> bool {
```

convention is `this`

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:372]: While I love `matches` and let-else, I think could be made more legible (and shorter) by just using let-chains: `if let … && let … { true } else { false }`.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:392]: However, that means we still accept `a[1, 2, 3]` and `a[0; 20]`, right? which isn't a valid expr.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:415]: Nice let-else!

File: tests/ui/parser/recover/recover-pat-exprs.stderr
Comment (fmease) [inline tests/ui/parser/recover/recover-pat-exprs.stderr]: Btw, if you'd like to suppress this one, we could pass a param `RecoverExprInPat::{Yes, No}` to `parse_pat…`

File: tests/ui/parser/recover/recover-pat-exprs.stderr
Comment (fmease) [inline tests/ui/parser/recover/recover-pat-exprs.stderr]: Note to self: Ah, right. If we were to reparse the tokens as an expr when failing to parse a pat (starting right before the first `(`), then we would only emit a single error for `(1 + 2) * 3` in pat position but we don't follow this approach since that would require creating a snapshot upfront potentially in the happy path very likely incurring a perf cost (it's been a while, trying to remember all the context).

File: tests/ui/parser/recover/recover-pat-exprs.rs
Comment (fmease) [inline tests/ui/parser/recover/recover-pat-exprs.rs:1]: Very nicely structured test file :clap:

**Review by fmease — COMMENTED** [review]
I've finished reviewing 3 out of 4 commits. Looking good so far. Gonna look at the 4th one in a sec. Sorry for taking so long ^^'

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:379]: This includes suffixed literals (e.g., `0i32`) unlike `is_two_tuple_indexes`. I'm fine either way as long as we're consistent which we're not right now.

File: compiler/rustc_parse/src/parser/expr.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/parser/expr.rs]: Yup, added a test.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/parser/pat.rs:403]: It uses `ExprKind::Dummy` as the left hand, so only [`IndexExpression`](https://doc.rust-lang.org/reference/expressions/array-expr.html#array-and-slice-indexing-expressions) is accepted, not [`ArrayExpression`](https://doc.rust-lang.org/reference/expressions/array-expr.html#array-expressions). Added a test.

File: compiler/rustc_parse/messages.ftl
Comment (fmease) [inline compiler/rustc_parse/messages.ftl:1]: Those suggestions are incredibly good! I didn't even think of suggesting match guards before!

Still, I think the wording is a bit too assertive atm given they are maybe-incorrect (we don't know if the recovered expr is actually const-evaluatable). I think we should use the wording `consider [verb]ing […]`  (e.g., `consider extracting the […]`) or something similar

File: tests/ui/parser/bad-name.stderr
Comment (fmease) [inline tests/ui/parser/bad-name.stderr:13]: I wonder if `let <expr>;` ever comes up in practice and if the suggestion is actually what the user actually intended. I guess it's fine

File: compiler/rustc_parse/messages.ftl
Comment (fmease) [reply compiler/rustc_parse/messages.ftl:783]: also maybe sth like *consider moving the expression to a match arm guard*

File: compiler/rustc_parse/src/errors.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/errors.rs]:
```
error: missing type for `const` item
 --> src/lib.rs:1:9
  |
1 | const X: /* Type */ = "foo";
  |         ^ help: provide a type for the constant: `&str`
  ```
Seems like it's already nicely handled :crab:

File: compiler/rustc_parse/src/errors.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/errors.rs]: Looks like [`print_expr_maybe_paren`](https://github.com/rust-lang/rust/blob/2ccafed862f6906707a390caf180449dd64cad2e/compiler/rustc_ast_pretty/src/pprust/state/expr.rs#L61), but the `Diagnostic` is already stashed by `maybe_recover_trailing_expr()`. I only have access to a `Diag` in `maybe_emit_stashed_expr_in_pats()`, not the recovered `ast::Expr` from the former `fn`.

File: compiler/rustc_errors/src/lib.rs
Comment (fmease) [reply compiler/rustc_errors/src/lib.rs:846]: @nnethercote Does this look okay to you?

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:469]: Since stashed diagnostics do get emitted eventually — even if not stolen (to be modified or replaced) — I think this method should be renamed not to contain `emit` but `maybe_modify` or sth. like that as well as the word `suggestion` maybe. This function essentially 'goes over' all expr-in-pat diagnostics and *maybe* enhances them by potentially adding statement/context-specific suggestions.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:496]: `modify_and_emit`?

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs]: In a future PR, we should probably make `CreateGuard` and `UpdateGuard` not use `expr` to achieve a 'cleaner'/more legible suggestion by not relying on `span_to_snippet`.

Rn, the suggestion copy/pastes the original expr into the suggestion:

```
3 |         val if val == 1 + 2 => {}
  |         ~~~ +++++++++++++++
```

if we just suggested only `val if val == ` instead, it would be

```
3 |         val if val == 1 + 2 => {}
  |         +++++++++++++
```

(and if we need to wrap the expr due to precedence, we can just use a multipart suggestion: `val if val == `, `)`)

It's always a win to rely less on `span_to_snippet` which is fallible (CC #128442, #128445).

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:602]: ... but yeah, I don't know how we can avoid calling `span_to_snippet` for `UnexpectedExpressionInPatternSugg::Const`

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:524]: Hmm, I'm not super convinced by this diagnostic. It has a lot of false positives. E.g., it suggests `1 + 2 = 3` for `let 1 + 2 = 3`. Which specific use case / user error did you want to target?

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:505]: Unwrapping `span_to_snippet` is pretty fishy but I can't think of a pathological example rn. Seems okay to keep for now.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:639]: Can be addressed some other time. I guess it's hard to tell what the user intended e.g. in `&mut &mut 1 + 2` (currently suggests `&mut const { &mut 1 + 2 }` etc.)

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [inline compiler/rustc_parse/src/parser/pat.rs:547]: That's unfortunate but it seems unavoidable

**Review by fmease — APPROVED** [review]
Thanks for your patience. Overall looks great! Ofc there's room for extra polish which can happen over time in follow-up PRs.

This does introduce a lot of complexity (stashing, stmt visiting in the parser, etc.) but I think these things are necessary / okay. In the future we might find some ways to simplify it but I really do like these diagnostics!

---

I've left some minor suggestions and notes. Most importantly:

1. I'm not convinced by the `RemoveLet` suggestion since it has a lot of false positives ([see here](https://github.com/rust-lang/rust/pull/123877#discussion_r1700569640)).
2. The `span_to_snippet(…).unwrap()` can actually lead to ICEs ([see here](https://github.com/rust-lang/rust/pull/123877#discussion_r1700635763))

After that we're pretty much good to go after fixup'ing and rebase ofc).

File: compiler/rustc_parse/src/parser/pat.rs
Comment (fmease) [reply compiler/rustc_parse/src/parser/pat.rs:505]: Actually, this can indeed ICE (similar to #128442, #128445 which I found after seeing this unwrap):

```rs
// a.rs
#[macro_export]
macro_rules! make {
    () => { match 0 { 0 + 0 => {} } }
}
```

```rs
// b.rs
fn main() { a::make!(); }
```

1. Compile `a`: `rustc +TOOLCHAIN a.rs --crate-type=lib`
2. Remove or move `a.rs`
3. Compile `b`: `rustc +TOOLCHAIN b.rs -L. --extern a --edition=2021`

---

Could you instead return early if `span_to_snippet` returns `None` and add a FIXME that `UnexpectedExpressionInPatternSugg::{Create,Update}Guard` don't need `span_to_snippet` / `expr : String` (as explained in another review comment of mine) meaning we don't need to return early for *them* (for `…Sugg::Const` it'll still be necessary tho).

File: compiler/rustc_errors/src/lib.rs
Comment (nnethercote) [reply compiler/rustc_errors/src/lib.rs:846]: Stashed diagnostics are tricky. Can you use one of the existing methods? Either `try_steal_modify_and_emit_err` or `try_steal_replace_and_emit_err`? I would strongly prefer that.

Comment (ShE3py) [reply compiler/rustc_errors/src/lib.rs]: Welp, it's unsound:
```rs
let guar = dcx.create_err(err).stash(span, key).unwrap();
dcx.steal_err(span, key, guar);
guar // `ErrorGuaranteed`, even tough it was cancelled
```
I'll look for an alternative.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/parser/pat.rs]: I opted for `maybe_augment_stashed_expr_in_pats_with_suggestions`

File: compiler/rustc_parse/src/parser/pat.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/parser/pat.rs]: I opted for `maybe_add_suggestions_then_emit`

File: compiler/rustc_parse/src/parser/pat.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/parser/pat.rs]: I tried to handle `let`s but yeah that doesn't make munch sense, I removed the suggestions for `let _;` & `let _ = ...;`.

I initially kept them for `let arr[i] = ...`, but I'm not sure if that's common.

File: compiler/rustc_errors/src/lib.rs
Comment (nnethercote) [reply compiler/rustc_errors/src/lib.rs]: I'm confused... what is unsound?

More generally, I'm very reluctant to muck around with the stashing API, adding new things to it.

Comment (ShE3py) [reply compiler/rustc_errors/src/lib.rs]: Ah mb, I thought `stash_diagnostic` created an `ErrorGuaranteed` out of thin air, but it actually creates a delayed bug.

Anyway, I'll look for a way that doesn't require modifying the stashing API.

Comment (nnethercote) [reply compiler/rustc_errors/src/lib.rs]: [This comment](https://github.com/rust-lang/rust/blob/9337f7afa6fda07e60d6aa5ee88e692878446782/compiler/rustc_errors/src/lib.rs#L782-L802) explains what `stash_diagnostic` does.

**Comment by fmease** [comment]
Btw, I will gladly insta-approve a PR that just contains the first two commits (a9fbb2276a583882872df6fad57d9dfcaebcc9f0 and f8b30463b1c663efdf321fc598e1a0799f6b93ca) which are independent of expr-in-pats :)

**Comment by ShE3py** [comment]
> Btw, I will gladly insta-approve a PR that just contains the first two commits ([a9fbb22](https://github.com/rust-lang/rust/commit/a9fbb2276a583882872df6fad57d9dfcaebcc9f0) and [f8b3046](https://github.com/rust-lang/rust/commit/f8b30463b1c663efdf321fc598e1a0799f6b93ca)) which are independent of expr-in-pats :)

I was on vacation and I wanted to move some comments (i.e. 2133219ddfeb1195a0aa3694026f7af8890f2f8d), I can split now if that's desirable.

---
I’ve reverted the “error reducing” for `parse_pat_tuple_or_parens` so as not to edit the stashing methods;
```rs
fn main() {
    let Some((1 + 2) * 3) = None::<i32>;
}
```
```
error: expected a pattern, found an expression
 --> src/main.rs:2:14
  |
2 |     let Some((1 + 2) * 3) = None::<i32>;
  |              ^^^^^^^^^^^ arbitrary expressions are not allowed in patterns

error: expected a pattern, found an expression
 --> src/main.rs:2:15
  |
2 |     let Some((1 + 2) * 3) = None::<i32>;
  |               ^^^^^ arbitrary expressions are not allowed in patterns
```
It's somewhat verbose, but that can be improved in a follow-up.

---
I wonder if the match arm guard suggestions should be moved to another PR? Tracking issue #129967 will move the guard from the match arm to the pattern, so it might be best to wait for `guard_patterns`'s implementation's PR to be merged before adding guard suggestions.

---
@rustbot ready

File: compiler/rustc_errors/src/lib.rs
Comment (nnethercote) [inline compiler/rustc_errors/src/lib.rs:1299]: I'd like to keep the stash API as minimal as possible. Please remove this and just do `stash`+`unwrap` at the call site. There are already a couple of call sites like that.

File: compiler/rustc_parse/src/parser/pat.rs
Comment (ShE3py) [reply compiler/rustc_parse/src/parser/pat.rs:472]: It requires the exact span which I don't have, and there's no
```rs
fn has_stashed_diagnostics(self: &DiagCtxtInner) -> bool {
    !self.stashed_diagnostics.is_empty()
}
```
So it's kinda the best I can do without modifying the stashing API.

**Review by fmease — APPROVED** [review]
> Tracking issue https://github.com/rust-lang/rust/issues/129967 will move the guard from the match arm to the pattern, so it might be best to wait for guard_patterns's implementation's PR to be merged before adding guard suggestions.

I see what you're getting at but erm, I'd rather merge this first. This PR is a lot older and will immediately affect diagnostics. We don't know how long it will take for the initial guard-pattern PR to get to a ready state and merged. Also, it shouldn't be that hard to update this code, I think.

I'll bors-approve after a rebase. Thanks a lot for your patience! I took way too long for this

**Pull request merged**
